doc/networking.md

```diff
@@ -107,22 +108,56 @@ The composition of the body (and certain headers) varies based on the content-ty
+4. Download the body
+5. Verify the keccak256 hash of the body matches `SegData.hash`
+6. Return 200 OK header. If any of the above steps fail, a non-200 response is returned.
+6. Transcode the segment
+7. Return a `TranscodeResult` body based on the results of the transcode.
 
 #### Response:
 
-200 OK after the upload has been processed. If the transcode subsequently fails, an  error message is returned. If there was an error in uploading or verifying the uploaded data, a 400- or 500- HTTP status code returned along with an error message. Otherwise, a success or error message is returned in the body.
+The response is split into two parts: the 200 OK  (or error) is sent after the download, and the response body consisting of a `TranscodeResult` is sent after the transcode completes. This gives broadcasters approximate visibility into how long the upload and transcode steps each take.
+
+```proto
+// Response that a transcoder sends after transcoding a segment.
+message TranscodeResult {
+
+    // Sequence number of the transcoded results.
+    int64 seq = 1;
+
+    // Result of transcoding can be an error, or successful with more info
+    oneof result {
+        string error = 2;
+        TranscodeData data = 3;
+    }
+}
+
+// A set of transcoded segments following the profiles specified in the job.
+message TranscodeData {
+
+    // Transcoded data, in the order specified in the job options
+    repeated TranscodedSegmentData segments = 1;
+
+    // Signature of the hash of the concatenated hashes
+    bytes sig = 2;
+}
+
+// Individual transcoded segment data.
+message TranscodedSegmentData {
+
+    // URL where the transcoded data can be downloaded from
+    string url = 1;
+
+}
+```
 
 ### Notes
 
-Currently, the stringified error is dumped directly into non-200 responses. This gives broadcasters more information to diagnose problems with remote transcoders. However, we may not want to return such details forever, as this may leak internal information that is best left private to a transcoder.
+Currently, any errors are dumped directly into the response in stringified form. This gives broadcasters more information to diagnose problems with remote transcoders. However, we may not want to return such details forever, as this may leak internal information that is best left private to a transcoder.
 
-Clients can use the difference in time between the request submissing and the 200ok to approximate the upload time. The time between the 200ok and receiving the response body approximates the transcode time.
+Broadcasters can use the difference in time between the request submissing and the 200ok to approximate the upload time. The time between the 200ok and receiving the response body approximates the transcode time.
 
 There is an end-to-end request timeout of 8 seconds. However, live streams where segments consistently take more than 4 seconds to process (the segment length) are likely to be outrun by players, leading to stuttering.
```

Comment:
Actually this is not entirely accurate - in case there transcoding takes more than 4 second transcoder will start to answer with `TranscoderBusy` error for new segments while previous one still transcoded - so player will see gaps in transcoded stream. But if transcode takes more than 8 seconds - there will be no segments in transcoded stream, because broadcaster will timeout and will not download transcoded segment.

Reply:
True that segments which take longer than 4 seconds to *transcode* will generally end up feeding back a `TranscoderBusy` error. This would also lead to stuttering on the player.

Note that "end to end" is inclusive of the entire request -- uploading, transcoding, etc. Suppose the *upload* consistently takes 4 seconds, but transcoding only takes 1 second; that's another case where playback outruns the end-to-end process. (Broadcasters would also have player/CDN uploads on top of this.) Being able to identify this issue is the reason for the separation between the 200ok and the response body. A `TranscodeBusy` message is one clue when debugging, and it's easy to see in the logs. I'm not sure other failure modes are as obvious, hence this note.

The key point here is that other problems will crop up before the 8 second timeout, any of which could impact playback.

In that respect, the fact that transcoders will only transcode 1 segment at a time, per job, is an implementation detail within the transcoder itself. That behavior should certainly be documented somewhere, but I'm not sure if the networking spec is the right place for it.

Also related: https://github.com/livepeer/go-livepeer/issues/491

Reply:
Fixed up the description a bit, let me know if it's better now